diff --git a/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs b/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
index 4394a8b..e32d338 100644
--- a/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
+++ b/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
@@ -36,12 +36,15 @@ namespace Microsoft.DotNet.Build.Tasks
         // should include the delimiter immediately before the RID, arch, or extension.
         protected string[] BadAtoms = new[] { "-x64", ".x64",
                                               ".tar", ".gz",
-                                              "-rhel.7", "-rhel.8",
-                                              ".rhel.7", ".rhel.8",
-                                              "-centos.7", "-centos.8",
-                                              ".centos.7", ".centos.8",
+                                              "-rhel.7", "-rhel.8", "-rhel.9",
+                                              ".rhel.7", ".rhel.8", ".rhel.9",
+                                              "-centos.7", "-centos.8", "-centos.9",
+                                              ".centos.7", ".centos.8", ".centos.9",
                                               ".fedora.30", "-fedora.30",
                                               ".fedora.31", "-fedora.31",
+                                              ".fedora.32", "-fedora.32",
+                                              ".fedora.33", "-fedora.33",
+                                              ".fedora.34", "-fedora.34",
                                               "-linux", ".linux",
                                               "-osx", ".osx",
                                               "-OSX", ".OSX",
