diff --git a/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs b/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
index 363fdeee..0a39a0d0 100644
--- a/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
+++ b/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/PublishCoreSetupBinaries.cs
@@ -46,6 +46,8 @@ namespace Microsoft.DotNet.Build.Tasks
                                               ".fedora.32", "-fedora.32",
                                               ".fedora.33", "-fedora.33",
                                               ".fedora.34", "-fedora.34",
+                                              ".fedora.35", "-fedora.35",
+                                              ".fedora.36", "-fedora.36",
                                               "-linux", ".linux",
                                               "-osx", ".osx",
                                               "-OSX", ".OSX",
